src/prefect/runner/runner.py

```diff
@@ -294,20 +319,45 @@ def goodbye_flow(name):
                     partial(
                         critical_service_loop,
                         workload=runner._check_for_cancelled_flow_runs,
-                        interval=self._query_seconds * 2,
+                        interval=self.query_seconds * 2,
                         run_once=run_once,
                         jitter_range=0.3,
                     )
                 )
 
-    def stop(self):
+    async def cancel_all(self):
+        runs_to_cancel = []
+
+        # done to avoid dictionary size changing during iteration
+        for flow_run_id, info in self._flow_run_process_map.items():
+            runs_to_cancel.append(info["flow_run"])
+        if runs_to_cancel:
+            for run in runs_to_cancel:
+                try:
+                    await self._cancel_run(run, state_msg="Runner is shutting down.")
+                except Exception:
+                    self._logger.exception(
+                        f"Exception encountered while cancelling {run.id}",
+                        exc_info=True,
+                    )
+
+    @sync_compatible
+    async def stop(self):
         """Stops the runner's polling cycle."""
         if not self.started:
             raise RuntimeError(
                 "Runner has not yet started. Please start the runner by calling"
                 " .start()"
             )
-        self._loops_task_group.cancel_scope.cancel()
+
+        self.stopping = True
+        await self.cancel_all()
```

Comment:
Doesn't need to be part of this PR, but it'd be nice to have a configurable grace period to wait before canceling on going runs.